Recognize <Find> and <Select> key codes

grawity opened the pull request:
These are the `ESC[1~` and `ESC[4~` keys that PuTTY sends for Home/End; they don't match what the 'xterm' terminfo has, so libtermkeys reports them as the original VT220 names.

I was hesitant to just permanently map them to K_HOME/K_END here, but having Neovim recognize them as their own keys will at the very least allow users to `:imap <Find> <Home>` or something along those lines.

(Of course, configuring PuTTY to set TERM=putty[-256color] would *in theory* be the preferred solution, but in practice it causes other problems due to other features missing from the 'putty' terminfo, as well as older systems that lack the terminfo entirely. Having Neovim recognize the keys is the least bad workaround.)

grawity committed:
fix(keycodes): recognize <Find> and <Select> as valid keys

PuTTY sets TERM=xterm, but sends ESC[1~ and ESC[4~ for Home/End keys,
which does not match what the 'xterm' terminfo has for khome/kend, so
libtermkeys instead reports them as the original DEC VT220 names.

The VT220 came with a DEC LK201 keyboard which had the following keys in
the area above arrow keys (where PCs now have Ins/Del/Home/End/etc):

  ┌────────┬────────┬────────┐
  │ Find   │ Insert │ Re-    │
  │        │ Here   │ move   │
  ├────────┼────────┼────────┤
  │ Select │ Prev   │ Next   │
  │        │ Screen │ Screen │
  └────────┴────────┴────────┘

These would send ESC[x~ sequences in the expected order:

  ┌────────┬────────┬────────┐
  │ ESC[1~ │ ESC[2~ │ ESC[3~ │
  ├────────┼────────┼────────┤
  │ ESC[4~ │ ESC[5~ │ ESC[6~ │
  └────────┴────────┴────────┘

Modern terminals continue to use the same sequences for Ins/Del as well
as PageUp/PageDn. But the VT220 keyboard apparently had no Home/End, and
PuTTY apparently chose to re-purpose the Find/Select key sequences for
Home/End (even though it claims to emulate Xterm and this doesn't match
what actual Xterm does).

So when Home/End are used in Neovim through PuTTY with TERM=xterm (the
default setting), libtermkey finds no match for the received sequences
in the terminfo database and defaults to reporting them as <Find> and
<Select> respectively.

PuTTY is not unique here -- tmux *also* sends ESC[1~ and ESC[4~ after
its internal translation -- but the difference is that 'tmux' terminfo
correctly maps them to Home/End so Neovim recognizes them as such, while
PuTTY defaults to using 'xterm' which uses a different mapping.

This initial patch only allows Neovim to recognize <Find> and <Select>
key codes as themselves, so that the user could manually map them e.g.
using ":imap <Find> <Home>".

Alternatives:

  - Using TERM=putty(-256color) would of course be the most correct
    solution, but in practice it leads to other minor issues, e.g. the
    need to have different PuTTY config profiles for older or non-Linux
    systems that lack that terminfo, or tmux's insistence on rendering
    italics as reverse.

  - Using Neovim through tmux avoids the problem (as tmux recognizes
    ESC[1~ on input), but is something that needs to be manually run
    every time.

The keycodes.h constants are slightly misnamed because K_SELECT was
already taken for a different purpose.

justinmk commented:
https://github.com/neovim/neovim/issues/2204 would be an alternative I guess, then you could `:set <F32>=^[...` and map to `<F32>` instead of e.g. `<Find>`. However the [latest comment](https://github.com/neovim/neovim/issues/2204#issuecomment-1890988865) there points out that this is not great for maintaining a config.

> I was hesitant to just permanently map them to K_HOME/K_END here

Is that because of the risk of "false positives"?

I can't think of any reason _not_ add more TERMCAP2KEY mappings, except that the `<Find>` and `<Select>` keycodes are a bit random, and Vim doesn't have them. Do you have a reference for the "DEC keys"?

Reference: last time we added keycodes: https://github.com/neovim/neovim/commit/d13803f64fc5607c6319087240e35a8b86082f64

If others don't have objections, this would need docs and a [test](https://github.com/neovim/neovim/blob/5e6240ffc24e55ecf7721fd9dc3f33c6f178be8c/test/functional/ui/input_spec.lua#L16).

zeertzjq commented on src/nvim/keycodes.h:
```diff
@@ -373,6 +373,9 @@ enum key_extra {
 #define K_KPAGEDOWN     TERMCAP2KEY('K', '5')   // keypad pagedown (lower R.)
 #define K_KORIGIN       TERMCAP2KEY('K', '2')   // keypad center
 
+#define K_KFIND         TERMCAP2KEY('@', '0')   // DEC keys ("Home" and "End"
+#define K_KSELECT       TERMCAP2KEY('*', '6')   // in PuTTY with TERM=xterm)
+
```
Why are these placed in the middle of keypad keys? I think these should be placed around `K_HELP`. And why are they named `K_KFIND` and `K_KSELECT` instead of `K_FIND` and `K_SELECT`?

grawity replied:
> Why are these placed in the middle of keypad keys?

I placed them there because that's where K_HOME and K_END are also defined, even though they're not keypad keys either.

> And why are they named K_KFIND and K_KSELECT instead of K_FIND and K_SELECT?

`K_SELECT` is already in use (by mouse-related things it seems) and I could not decide what other name to choose. Renaming the existing K_SELECT seemed too invasive so I thought "I'll name it K_KSELECT until someone suggests better names".

zeertzjq replied:
Oh, then I think `K_KFIND` and `K_SELECT` are fine.

zeertzjq commented on src/nvim/keycodes.c:
```diff
@@ -283,6 +283,9 @@ static const struct key_name_entry {
   { K_KORIGIN,         "kOrigin" },
   { K_KORIGIN,         "KP5" },
 
+  { K_KFIND,           "Find" },        // "Home" and "End" in
+  { K_KSELECT,         "Select" },      // PuTTY with TERM=xterm
+
```
ditto

grawity commented:
> except that the `<Find>` and `<Select>` keycodes are a bit random

I wasn't the one who made them up. Those are the existing key names that neovim *already* receives from libtermkey in this situation (as I've found out via strace between the two nvim processes – I literally saw the strings `<Find>` and `<Select>` being exchanged through the RPC).

But currently Neovim completely ignores them and doesn't allow mapping them to anything else; the patch just makes that possible.

> Do you have a reference for the "DEC keys"?

My assumption is that they're named after the DEC LK201 keyboard that the physical VT220 used. Looking at photos online, it had 'Find' as key 1 and 'Select' as key 4 in roughly the same area where we now have Ins/Home/Del/End, so it seems to fit the explanation. I didn't dig deeper into why libtermkey uses those names.

justinmk commented:
@grawity thanks, that is great info to add to the commit message!

grawity commented:
> Last time we added keycodes: [d13803f](https://github.com/neovim/neovim/commit/d13803f64fc5607c6319087240e35a8b86082f64)

Okay, so looking at this, I assume I could add `case K_FIND: ... return VTERM_KEY_HOME` in terminal.c to permanently?

After thinking more about it, false positives are probably unlikely, as tmux also sends the same key sequences when translating Home/End (and PC keyboards do not have Find/Select keys anyway).

I moved the keycode definitions next to K_HELP and renamed K_KFIND to K_FIND, but still not sure what better name to use instead of K_KSELECT (as K_SELECT is already taken).

leonerd commented:
@grawity You're right - I picked the names to match what the original DEC keyboard and terminals would do there. A fair number of modern terminals specifically use "CSI H" and "CSI F" as encodings for the `<Home>` and `<End>` keys, rather than functions 1 and 4.
I'm not really familiar with how neovim uses libvterm or not, and therefore what the logic of the TERMCAP2KEY function is doing here, so I'm not sure how much further I could help here.

justinmk commented:
@grawity I missed this somehow. Please rebase then I'll merge it.
